_posts/18-05-2022-a_story_unfolds.md

```diff
@@ -0,0 +1,223 @@
+
+Let's try to imagine a function like this. First, we need to be able to look around:
+```haskell
+type LookAround a = a -> [a]
+```
+Then, we need to be able to step along the way and construct subtrees - but this is simple recursion. We finally need to be able to construct our node (given the value to store inside the node and the paths to follow):
+```haskell
+type BuildNode t a = a -> [t a] -> t a
+```
+Having these, we can construct our tree builder:
+```haskell
+treeBuilder :: LookAround a -> BuildNode t a -> a -> t a
+treeBuilder lookAround buildNode value = (buildNode value . childNodes . lookAround) value
+  where childNodes = fmap (treeBuilder lookAround buildNode)
+```
+To be more confident that it works as expected, we can, for example, define these:
+```haskell
+data Tree a = Leaf a | Node a [Tree a] deriving Show
+
+children :: LookAround Int
+children 1 = []
+children n = filter (\num -> n `mod` num == 0) [1..n-1]
+
+nodes :: BuildNode Tree Int
+nodes value [] = Leaf value
+nodes value subtrees = Node value subtrees
+```
+
+and then run:
+```haskell
+print (treeBuilder children nodes 12)
+```
+to get a tree in which children of the node are the divisors of the value `v` stored in the node (smaller than `v` itself), `12` being the root.
+
+```
+Node 12 [
+  Leaf 1,
+  Node 2 [Leaf 1],
+  Node 3 [Leaf 1],
+  Node 4 [ 
+    Leaf 1,
+    Node 2 [Leaf 1]
+  ],
+  Node 6 [
+    Leaf 1,
+    Node 2 [Leaf 1],
+    Node 3 [Leaf 1] 
+  ]
+]
+```
+It is neither useful nor spectacular, but gives us some reassurance. The only remaining problem is, how does `IO` fit the puzzle? Obviously, the `treeBuilder`, albeit nice, cannot handle the side effects.
+
+So after another brief moment of hacking, I managed to stitch together an implementation:
+```haskell
+type LookAroundM m a = a -> m [a]
+type BuildNodeM t m a = a -> [t a] -> m (t a)
+
+treeBuilderM :: Monad m => LookAroundM m a -> BuildNodeM t m a -> a -> m (t a)
+treeBuilderM lookAroundM buildNodeM value = lookAroundM value >>= childNodes >>= buildNodeM value
+  where childNodes = traverse (treeBuilderM lookAroundM buildNodeM)
+```
+Now it contains the scary M-word, but it lets us redefine `treeBuilder` as a one-liner:
+```haskell
+treeBuilder lookAround buildNode value = runIdentity (treeBuilderM (return . lookAround) (\a b -> return (buildNode a b)) value)
+```
+The `lookAround` function now looks like this:
+```haskell
+lookAround :: LookAroundM IO File
+lookAround file = do
+  isDirectory <- isDir file
+  if isDirectory
+  then listDir file
+  else return []
+```
+and the `buildNode` becomes:
+```haskell
+buildNode :: File -> [FileTree File] -> IO (FileTree File)
+buildNode file subtrees = do
+  isDirectory <- isDir file
+  if isDirectory
+  then return (Directory file subtrees)
+  else return (RegularFile file)
+```
+with a minor modification to our `FileTree`:
+```haskell
+data FileTree a = RegularFile a | Directory a [FileTree a]
```

Comment:
Huh `FileTree` is now type-parametrized (or whatever it's called in Haskell)... but is it ever used with anything else than `File` as type param? 🤔

Reply:
Right, at this point it would be better to use the more generic `Tree` type.

Reply:
Yeah, parametrizing a `Tree` with `File would make sense (unlike param'zing `FileTree` with `File`)